server.go

```diff
@@ -715,10 +716,14 @@ func (s *Server) processPublish(cl *Client, pk packets.Packet) error {
 		pk.FixedHeader.Qos = s.Options.Capabilities.MaximumQos // [MQTT-3.2.2-9] Reduce Qos based on server max qos capability
 	}
 
-	if pkx, err := s.hooks.OnPublish(cl, pk); err == nil {
+	pkx, err := s.hooks.OnPublish(cl, pk)
+	if err == nil {
 		pk = pkx
 	} else if errors.Is(err, packets.ErrRejectPacket) {
```

Comment:
Can we add a new internal Error code (Ignore) here. Basically it will not further publish to the subscribers, but it will build the proper acks. Currently, ErrRejectPacket works like that only for QoS 0.

packets.ErrRejectPacket can be captured in the next else if.

Reply:
> Currently, ErrRejectPacket works like that only for QoS 0.

I'm not sure this is the case because
```go
} else if errors.Is(err, packets.ErrRejectPacket) {
    return nil
```
will return before `s.publishToSubscribers(pk)` occurs 10 lines later. As it stands, RejectPacket will effectively behave if the packet never existed.

We could add a `packets.ErrIgnorePacket` as you say, which will simply not retain or publish out the incoming message, but pretend that the packet was received normally. Is this what you mean?

Reply:
Yes. Ignore will pretend that packet is received normally, but will not retain or publish.